### PR TITLE
Initial version of perforce-helix-core-server template.

### PR DESCRIPTION
**Pull request by rcowham**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [ x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

* Initial version of template

### CODE REVIEW

**Comment by rcowham** [comment]
I can see errors in validation:
`    [-] Variables Must Be Referenced (103 ms) 
##[error]        Cannot bind argument to parameter 'Name' because it is an empty 
##[error]        Unreferenced variable: 
##[error]        Cannot bind argument to parameter 'Name' because it is an empty 
##[error]        Unreferenced variable: 
##[error]        Cannot bind argument to parameter 'Name' because it is an empty 
##[error]        Unreferenced variable: 
`
Is this an error in the validation script (eg. the regex in use)?

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+            "keyData": "[parameters('adminSSHPubKey')]"
+          }
+        ]
+      }
+    }
+  },
+  "resources": [
+    {
+      "apiVersion": "2019-05-01",
+      "name": "pid-6c922c23-cc65-4b76-8290-74feb0f3c476",
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:179]: is this registered?

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+      },
+      "RHEL 7.x": {
+        "publisher": "RedHat",
+        "offer": "RHEL",
+        "sku": "7-LVM",
+        "version": "latest"
+      }
+    },
+    "dataDiskSize": "[parameters('dataDiskSize')]",
+    "publicIpName": "[substring(concat(uniqueString(resourceGroup().id, deployment().name)), 0, 6)]",
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:118]: make this idempotent (e.g. remove deployment().name) - also remove extraneous concat()

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+          "description": "The base URI where artifacts required by this template are located including a trailing '/'"
+      },
+      "defaultValue": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-perforce-helix-core-server/"
+    },
+    "_artifactsLocationSasToken": {
+      "type": "securestring",
+      "metadata": {
+          "description": "The sasToken required to access _artifactsLocation.  When the template is deployed using the accompanying scripts, a sasToken will be automatically generated. Use the defaultValue if the staging location is not secured."
+      },
+      "defaultValue": ""
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:17]: move all the defaultValues above metadata

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    "location": {
+      "type": "string",
+      "metadata": {
+        "description": "The location where these resources will be deployed."
+      },
+      "defaultValue": "[resourceGroup().location]"
+    },
+    "OS": {
+      "type": "string",
+      "allowedValues": ["Ubuntu 18.04 LTS", "CentOS 7.5", "RHEL 7.x"],
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:28]: format the JSON (e.g. SHIFT+ALT+F in VS Code)

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    "dataDiskSize": {
+      "type": "int",
+      "minValue": 10,
+      "maxValue": 2048,
+      "metadata": {
+        "description": "Please select the size of the data disk you wish to deploy (value is integer GB) to hold your metadata, logs and depot (archive files). This can be any value up to 2TB (2048 GB)."
+      },
+      "defaultValue": 50
+    },
+    "p4Port": {
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:50]: add defaultValues for the ports

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+        "description": "Admin username for Virtual Machine"
+      }
+    },
+    "adminSSHPubKey": {
+      "type": "securestring",
+      "metadata": {
+        "description": "SSH Public Key for the Virtual Machine."
+      }
+    },
+    "source_CIDR": {
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:74]: add an option (and default for "*")

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        "storageProfile": {
+          "imageReference": "[variables('imageReference')[parameters('OS')]]",
+          "osDisk": {
+            "caching": "ReadWrite",
+            "createOption": "FromImage"
+          },
+          "copy": [
+            {
+              "name": "dataDisks",
+              "count": 1,
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:308]: if the only option is a single disk, remove the copy loop

File: 101-perforce-helix-core-server/azuredeploy.parameters.json [hunk 8]
```diff
@@ -0,0 +1,33 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+  "parameters": {
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.parameters.json:4]: remove the params you have defaultValues for

File: 101-perforce-helix-core-server/azuredeploy.json (hunk 1 above)
Comment (rcowham) [reply 101-perforce-helix-core-server/azuredeploy.json:179]: I just copied that field. Have changed it to a simple string name: "helix_core_server". That OK?

File: 101-perforce-helix-core-server/azuredeploy.json (hunk 6 above)
Comment (rcowham) [reply 101-perforce-helix-core-server/azuredeploy.json:74]: Not sure what you mean by 'option'. Have added default.

**Comment by rcowham** [comment]
@bmoore-msft Are you able to merge this now?

File: 101-perforce-helix-core-server/azuredeploy.json [hunk 9]
```diff
@@ -175,15 +183,16 @@
   },
   "resources": [
     {
-      "apiVersion": "2019-05-01",
-      "name": "pid-6c922c23-cc65-4b76-8290-74feb0f3c476",
+      "apiVersion": "2019-10-01",
```
Comment (bmoore-msft) [inline 101-perforce-helix-core-server/azuredeploy.json:186]: this is for marketplace publisher identification, so just remove it from the sample

Comment (rcowham) [reply 101-perforce-helix-core-server/azuredeploy.json]: OK - removed the resource.

**Comment by rcowham** [comment]
@bmoore-msft Hopefully that addresses things?

**Pull request merged**